llvm/lib/IR/ConstantRange.cpp

```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
   return UR.isSizeStrictlySmallerThan(SR) ? UR : SR;
 }
 
+ConstantRange
+ConstantRange::multiplyWithNoWrap(const ConstantRange &Other,
+                                  unsigned NoWrapKind,
+                                  PreferredRangeType RangeType) const {
+  if (isEmptySet() || Other.isEmptySet())
+    return getEmpty();
+
+  ConstantRange Result = multiply(Other);
+
+  // TODO: Bounds are not yet precise if the input sets are non-wrapped /
```

Comment:
Could you please provide a negative test or an example? It's hard for me to imagine it.

Reply:
For example, given input ranges `[0,2)` and `[2,0)` (bitwidth: 4), we currently return `full-set`, but a more refined CR would be `[2,1)`.